### PR TITLE
remove kvs from ttxdb

### PR DESCRIPTION
**Pull request by adecaro**

### CODE REVIEW

File: token/services/ttxdb/db/badger/badger.go [hunk 1]
```diff
@@ -512,6 +521,91 @@ func (db *Persistence) GetStatus(txID string) (driver.TxStatus, error) {
...
+
+	txn := db.db.NewTransaction(true)
+	defer txn.Discard()
+	err = txn.Set([]byte(key), b)
+	if err != nil {
+		return errors.Wrapf(err, "could not set value for key %s", key)
+	}
+	err = txn.Commit()
+	if err != nil {
+		return errors.Wrapf(err, "could not committ transaction for [%s]", txID)
```
Comment (arner) [inline token/services/ttxdb/db/badger/badger.go:550]: committ = typo

File: token/services/ttxdb/db/badger/badger.go [hunk 2]
```diff
@@ -512,6 +521,91 @@ func (db *Persistence) GetStatus(txID string) (driver.TxStatus, error) {
 	return driver.Unknown, errors.Errorf("transaction [%s] not found", txID)
 }
 
+func (db *Persistence) AddTransactionEndorsementAck(txID string, id view.Identity, sigma []byte) error {
+	key, err := db.transactionEndorseAckKey(txID, id)
+	if err != nil {
+		return errors.Wrapf(err, "could not get key for token request %s", txID)
+	}
+	logger.Debugf("Adding token request [%s] with key [%s]", txID, key)
```
Comment (arner) [inline token/services/ttxdb/db/badger/badger.go:529]: I think we start logs with lowercase?

File: token/services/ttxdb/db/sql/sql.go [hunk 3]
```diff
@@ -329,6 +330,63 @@ func (db *Persistence) QueryValidations(params driver.QueryValidationRecordsPara
...
+				logger.Warnf("tried to get status for non-existent tx %s, returning unknown", txID)
+				continue
+			}
+			return nil, errors.Wrapf(err, "error querying db")
+		}
+		fmt.Printf(" found [%s] [%s]\n", id, sigma)
+		acks[view.Identity(id).String()] = sigma
+	}
+
+	fmt.Printf(" result [%v]\n", acks)
```
Comment (arner) [inline token/services/ttxdb/db/sql/sql.go:386]: remove this (or change to debug log)

File: token/services/ttxdb/db/sql/sql.go [hunk 4]
```diff
@@ -329,6 +330,63 @@ func (db *Persistence) QueryValidations(params driver.QueryValidationRecordsPara
...
+		var sigma []byte
+		if err := rows.Scan(&id, &sigma); err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				// not an error for compatibility with badger.
+				logger.Warnf("tried to get status for non-existent tx %s, returning unknown", txID)
+				continue
+			}
+			return nil, errors.Wrapf(err, "error querying db")
+		}
+		fmt.Printf(" found [%s] [%s]\n", id, sigma)
```
Comment (arner) [inline token/services/ttxdb/db/sql/sql.go:382]: Remove or change to debug log

File: token/services/ttxdb/db/sql/random.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/pkg/errors"
+)
+
+const (
+	// NonceSize is the default NonceSize
+	NonceSize = 24
+)
+
+// GetRandomBytes returns len random looking bytes
+func GetRandomBytes(len int) ([]byte, error) {
```
Comment (arner) [inline token/services/ttxdb/db/sql/random.go:31]: In other places, the SQL driver uses go-uuid to generate a unique ID (see https://github.com/hyperledger-labs/fabric-token-sdk/blob/ad5e98882ce2cad448e1e4f397f912e416203552/token/services/ttxdb/db/sql/sql.go#L177). I propose to stick to that.

Comment (adecaro) [reply token/services/ttxdb/db/sql/random.go:31]: Perfect, I'll use that.

**Comment by adecaro** [comment]
@arner thanks. I have addressed all your comments :)

**Review by arner — APPROVED** [review]

File: token/services/ttxdb/db/badger/badger.go [hunk 6]
```diff
@@ -512,6 +521,91 @@ func (db *Persistence) GetStatus(txID string) (driver.TxStatus, error) {
...
+	type Entry struct {
+		key   string
+		value []byte
+	}
+	var entries []Entry
+	stream := db.db.NewStream()
+	stream.NumGo = db.numGoStream
+	stream.LogPrefix = streamLogPrefixStatus
+	txIdAsBytes := []byte(txID)
+	stream.ChooseKey = func(item *badger.Item) bool {
```
Comment (alexandrosfilios) [inline token/services/ttxdb/db/badger/badger.go:569]: If I understand correctly, you only choose the "tea"-prefixed entries later. In this case I see that badger has a `Prefix` property that I think filters out more efficiently. This way you skip the check later and you iterate over less keys.

File: token/services/ttxdb/db/badger/badger.go [hunk 7]
```diff
@@ -512,6 +521,91 @@ func (db *Persistence) GetStatus(txID string) (driver.TxStatus, error) {
...
+
+func (db *Persistence) GetTransactionEndorsementAcks(txID string) (map[string][]byte, error) {
+	acks := make(map[string][]byte)
+
+	// search for all matching keys
+	type Entry struct {
+		key   string
+		value []byte
+	}
+	var entries []Entry
```
Comment (alexandrosfilios) [inline token/services/ttxdb/db/badger/badger.go:564]: I see that another method `SetStatus()` and `GetStatus()` use the same logic for the streams. Maybe you could extract this functionality into a method:
```
func (db *Persistence) GetEntriesByTxID(txID string, prefix string) []Entries {
 var entries []Entry
	stream := db.db.NewStream()
	stream.NumGo = db.numGoStream
	stream.LogPrefix = streamLogPrefixStatus
	txIdAsBytes := []byte(txID)
	stream.ChooseKey = func(item *badger.Item) bool {
		return bytes.HasSuffix(item.Key(), txIdAsBytes)
	}
if len(prefix) > 0 { stream.Prefix = []byte(prefix)}
	stream.Send = func(buf *z.Buffer) error {
		list, err := badger.BufferToKVList(buf)
		if err != nil {
			return err
		}
		for _, kv := range list.Kv {
			entries = append(entries, Entry{key: string(kv.Key), value: kv.Value})
		}
		return nil

	}
	if err := stream.Orchestrate(context.Background()); err != nil {
		return err
	}
return entries
}
```

File: token/services/ttxdb/db/badger/marshaller.go [hunk 8]
```diff
@@ -28,6 +28,11 @@ func MarshalTokenRequest(mr *TokenRequest) ([]byte, error) {
 	return json.Marshal(mr)
 }
 
+// MarshalTransactionEndorseAck marshals a TransactionEndorseAck into a byte array
```
Comment (alexandrosfilios) [inline token/services/ttxdb/db/badger/marshaller.go:31]: The wrapper methods for marshalling make it a bit harder to re-use code. Do we need to have them separate?

File: token/services/ttxdb/db/badger/badger.go [hunk 9]
```diff
@@ -512,6 +521,91 @@ func (db *Persistence) GetStatus(txID string) (driver.TxStatus, error) {
...
+
+	if len(entries) == 0 {
+		// nothing to update
+		logger.Debugf("no entries found for txID %s, skipping", txID)
+		return nil, nil
+	}
+
+	for _, entry := range entries {
+		switch {
+		case strings.HasPrefix(entry.key, "tea"):
```
Comment (alexandrosfilios) [inline token/services/ttxdb/db/badger/badger.go:596]: Maybe we could have a string-backed enum for these prefixes

**Pull request merged**